### PR TITLE
config: add snap config model

### PR DESCRIPTION
**Pull request by mr-cal**
- [X] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [X] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [X] Have you successfully run `make lint`?
- [X] Have you successfully run `pytest tests/unit`?

-----
Allows Snapcraft to parse data set via `snap set snapcraft <key>=<value>`

Currently, the only supported parameter is `provider`, which will be used in an upcoming PR to allow users to choose a default provider.

(CRAFT-1321)

### CODE REVIEW

File: snap/hooks/configure [hunk 1]
```diff
@@ -0,0 +1,11 @@
+#! /bin/bash
+
+config_file="$SNAP_DATA/snap-config.yaml"
+provider="$(snapctl get provider)"
+
+cat <<EOT > "${config_file}"
+# This file is generated automatically generated by snapcraft's configure hook.
```
Comment (sergiusens) [inline snap/hooks/configure:7]: double generated

File: snap/hooks/configure [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#! /bin/bash
+
+config_file="$SNAP_DATA/snap-config.yaml"
+provider="$(snapctl get provider)"
+
+cat <<EOT > "${config_file}"
```
Comment (sergiusens) [inline snap/hooks/configure:6]: instead of writing to a file why don't we call get in the relevant locations?

Comment (mr-cal) [reply snap/hooks/configure:6]: Because `snap get` requires root access

File: snap/hooks/configure (hunk 1 above)
Comment (mr-cal) [reply snap/hooks/configure]: Fixed

File: snap/hooks/configure (hunk 2 above)
Comment (mr-cal) [reply snap/hooks/configure:6]: Details for posterity:
### snap get
- always requires root access
### snapctl get
- doesn't require root access when executed inside a snap
- fails when executed outside a snap (like during a unit test...)

**Pull request closed without merging**